src/main/java/com/epam/esm/service/CertificateService.java

```diff
@@ -0,0 +1,205 @@
+                return new ResponseEntity<>(savedGiftCertificate, HttpStatus.CREATED);
+
+        } else {
+            GiftCertificate foundCertificate = tryToFindCertificate.get();
+            Long idFound = foundCertificate.getId();
+            ErrorDTO errorResponse = new ErrorDTO(CERTIFICATE_ALREADY_EXISTS.formatted(idFound), CERTIFICATE_FOUND);
+            return ResponseEntity.status(HttpStatus.FOUND).body(errorResponse);
+        }
+    }
+
+    public ResponseEntity<?> getGiftCertificateById(@NonNull Long giftCertificateId) {
+
+        if (certificateRepository.existsById(giftCertificateId)) {
+            Optional<GiftCertificate> giftCertificate = certificateRepository.findById(giftCertificateId);
+            return ResponseEntity.status(HttpStatus.FOUND).body(giftCertificate);
+        } else {
+            String message = CERTIFICATE_WITH_ID_NOT_FOUND.formatted(giftCertificateId);
+
+            ErrorDTO errorResponse = new ErrorDTO(message, CERTIFICATE_NOT_FOUND);
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorResponse);
+        }
+    }
+
+    public ResponseEntity<List<GiftCertificate>> getFilteredCertificates(
+        String tagName, String searchWord, String nameOrder, String createDateOrder) {
+
+        Specification<GiftCertificate> spec = new CertificateSpecification(tagName, searchWord);
+        Sort sort = constructSort(nameOrder, createDateOrder);
+
+        List<GiftCertificate> certificates = certificateRepository.findAll(spec, sort);
+
+        return ResponseEntity.status(HttpStatus.OK).body(certificates);
+    }
+    private Sort constructSort(String nameOrder, String createDateOrder) {
+        Sort sortBy = Sort.unsorted();
+
+        if ("ASC".equalsIgnoreCase(nameOrder)) {
+            sortBy = sortBy.and(Sort.by("name").ascending());
+        } else if ("DESC".equalsIgnoreCase(nameOrder)) {
+            sortBy = sortBy.and(Sort.by("name").descending());
+        }
+
+        if ("ASC".equalsIgnoreCase(createDateOrder)) {
+            sortBy = sortBy.and(Sort.by("createDate").ascending());
+        } else if ("DESC".equalsIgnoreCase(createDateOrder)) {
+            sortBy = sortBy.and(Sort.by("createDate").descending());
+        }
+
+        return sortBy;
+    }
+
+    public ResponseEntity<?> deleteGiftCertificate(Long giftCertificateId) {
+
+        if(certificateRepository.existsById(giftCertificateId)){
+            certificateRepository.deleteById(giftCertificateId);
+            return ResponseEntity.status(HttpStatus.FOUND).body(null);
+        }
+        else
+            return ResponseEntity.status(HttpStatus.NO_CONTENT).body(
+            new ErrorDTO(
+                CERTIFICATE_WITH_ID_NOT_FOUND.formatted(giftCertificateId),
+                CERTIFICATE_NOT_FOUND));
+    }
+
+    @Transactional
+    public GiftCertificate updateGiftCertificate(@NonNull Long certificateId, GiftCertificate updates, List<Long> newTagIdsList) {
```

Comment:
This method differs from others:  in others you construct response and here you just return certificate. Consider refactoring it to also return  ResponseEntity<?> (would likely requiring moving some logic from controller to service).

Reply:
Changed